exchanges/binance/ratelimit.go

```diff
@@ -29,18 +43,84 @@ type RateLimit struct {
 
 // Limit executes rate limiting functionality for Binance
 func (r *RateLimit) Limit(f request.EndpointLimit) error {
-	if f == request.Auth {
-		time.Sleep(r.Orders.Reserve().Delay())
-		return nil
+	var limiter *rate.Limiter
+	var tokens int
+	switch f {
+	case limitHistoricalTrades:
+		limiter, tokens = r.GlobalRate, 5
+	case limitOrderbookDepth500:
+		limiter, tokens = r.GlobalRate, 5
+	case limitOrderbookDepth1000:
+		limiter, tokens = r.GlobalRate, 10
+	case limitOrderbookDepth5000:
+		limiter, tokens = r.GlobalRate, 50
+	case limitOrderbookTickerAll:
+		limiter, tokens = r.GlobalRate, 2
+	case limitPriceChangeAll:
+		limiter, tokens = r.GlobalRate, 40
+	case limitSymbolPriceAll:
+		limiter, tokens = r.GlobalRate, 2
+	case limitOpenOrdersAll:
+		limiter, tokens = r.Orders, 40
+	case limitOrder:
+		limiter, tokens = r.Orders, 1
+	case limitOrdersAll:
+		limiter, tokens = r.Orders, 5
+	default:
+		limiter, tokens = r.GlobalRate, 1
 	}
-	time.Sleep(r.GlobalRate.Reserve().Delay())
+
+	var finalDelay time.Duration
+	for i := 0; i < tokens; i++ {
+		// Consume tokens 1 at a time as this avoids needing burst capacity in the limiter,
+		// which would otherwise allow the rate limit to be exceeded over short periods
+		finalDelay = limiter.Reserve().Delay()
+	}
+	time.Sleep(finalDelay)
 	return nil
 }
 
 // SetRateLimit returns the rate limit for the exchange
 func SetRateLimit() *RateLimit {
 	return &RateLimit{
-		GlobalRate: request.NewRateLimit(binanceGlobalInterval, binanceOrderDailyMaxRequests),
+		GlobalRate: request.NewRateLimit(binanceGlobalInterval, binanceGlobalRequestRate),
 		Orders:     request.NewRateLimit(binanceOrderInterval, binanceOrderRequestRate),
 	}
 }
+
+func bestPriceLimit(symbol string) request.EndpointLimit {
+	if symbol == "" {
+		return limitOrderbookTickerAll
+	}
+
+	return limitDefault
+}
+
+func openOrdersLimit(symbol string) request.EndpointLimit {
+	if symbol == "" {
+		return limitOpenOrdersAll
+	}
+
+	return limitOrder
+}
+
+func orderbookLimit(depth int) request.EndpointLimit {
```

Comment:
Please add test coverage for these (even though they're self explanatory)

Reply:
Included tests for each of the helper cases, and invoking the `Limit` function with each return value (on a brand new instance so there is no stacking limits). This does mean that the tests have to wait for the full delay to execute, which is worst in the case of 'all open orders' (40 tokens required, with 10/sec limit)

Reply:
Love your work @dackroyd!